tempesta_fw/sock_clnt.c

```diff
@@ -164,12 +164,14 @@ tfw_sock_clnt_new(struct sock *sk)
 	}
 
 	/* Link Tempesta with the socket and the peer. */
-	tfw_connection_revive(conn);
 	tfw_connection_link_to_sk(conn, sk);
 	tfw_connection_link_from_sk(conn, sk);
 	tfw_connection_link_peer(conn, (TfwPeer *)cli);
 	ss_set_callbacks(sk);
 
+	/* The connection is available to others now. */
+	tfw_connection_revive(conn);
```

Comment:
Why did you move the refcounter initialization? When you "publish" connection to the world anything can happen with it, so we can go to improper manipulations with uninitialized refcounter.